src/Language/PureScript/TypeChecker/Monad.hs

```diff
@@ -200,16 +201,17 @@ modifyEnv f = modify (\s -> s { checkEnv = f (checkEnv s) })
 -- |
 -- Run a computation in the Check monad, starting with an empty @Environment@
 --
-runCheck :: (MonadError MultipleErrors m) => Check a -> m (a, Environment)
+runCheck :: (MonadError MultipleErrors m, MonadWriter MultipleErrors m) => Check a -> m (a, Environment, MultipleErrors)
 runCheck = runCheck' initEnvironment
 
 -- |
 -- Run a computation in the Check monad, failing with an error, or succeeding with a return value and the final @Environment@.
 --
-runCheck' :: (MonadError MultipleErrors m) => Environment -> Check a -> m (a, Environment)
-runCheck' env c = interpretMultipleErrors $ do
-  (a, s) <- flip runStateT (CheckState env 0 0 Nothing) $ unCheck c
-  return (a, checkEnv s)
+runCheck' :: (MonadError MultipleErrors m, MonadWriter MultipleErrors m) => Environment -> Check a -> m (a, Environment, MultipleErrors)
```

Comment:
Here's the goal: we want to unwrap all the constructors of the monad transformers making up `Check` and get some data structure. It's something involving `Either`, `(,)`, `Environment`, `MultipleErrors` and `CheckState`.

Then we want to pass that data structure to `interpretMultipleErrors` to interpret it in some arbitrary `Monad` with the right `mtl` constraints. 

Try breaking the problem down into those two parts. Write `runCheck' env = interpretMultipleErrors . unwrapCheck env`, and try to implement `unwrapCheck` in a `where` clause. Then, you need to modify `interpretMultipleErrors` to handle warnings. Right now it only interprets errors. I would also rename it, to something like `interpretErrorsAndWarnings`.